libmogwai-schedule/tests/signal-logger.h

```diff
@@ -0,0 +1,140 @@
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
+ *
+ * Authors:
+ *  - Philip Withnall <withnall@endlessm.com>
+ */
+
+#pragma once
+
+#include <glib.h>
+#include <glib-object.h>
+
+G_BEGIN_DECLS
+
+typedef struct _MwsSignalLoggerEmission MwsSignalLoggerEmission;
+
+void             mws_signal_logger_emission_free       (MwsSignalLoggerEmission *self);
+void             mws_signal_logger_emission_get_params (MwsSignalLoggerEmission *self,
+                                                        ...);
+
+G_DEFINE_AUTOPTR_CLEANUP_FUNC (MwsSignalLoggerEmission, mws_signal_logger_emission_free)
+
+typedef struct _MwsSignalLogger MwsSignalLogger;
+
+MwsSignalLogger *mws_signal_logger_new     (void);
+void             mws_signal_logger_free    (MwsSignalLogger *self);
+gulong           mws_signal_logger_connect (MwsSignalLogger *self,
+                                            gpointer         obj,
+                                            const gchar     *signal_name);
+
+G_DEFINE_AUTOPTR_CLEANUP_FUNC (MwsSignalLogger, mws_signal_logger_free)
+
+gsize            mws_signal_logger_get_n_emissions  (MwsSignalLogger                *self);
+gboolean         mws_signal_logger_pop_emission     (MwsSignalLogger                *self,
+                                                     gpointer                       *out_obj,
+                                                     gchar                         **out_obj_type_name,
+                                                     gchar                         **out_signal_name,
+                                                     MwsSignalLoggerEmission       **out_emission);
+gchar           *mws_signal_logger_format_emission  (gpointer                        obj,
+                                                     const gchar                    *obj_type_name,
+                                                     const gchar                    *signal_name,
+                                                     const MwsSignalLoggerEmission  *emission);
+gchar           *mws_signal_logger_format_emissions (MwsSignalLogger                *self);
+
+/**
+ * mws_signal_logger_assert_no_emissions:
+ * @self: a #MwsSignalLogger
+ *
+ * Assert that there are no signal emissions currently in the logged stack.
+ *
+ * Since: 0.1.0
+ */
+#define mws_signal_logger_assert_no_emissions(self) \
+  G_STMT_START { \
+    if (mws_signal_logger_get_n_emissions (self) > 0) \
+      { \
+        g_autofree gchar *ane_list = mws_signal_logger_format_emissions (self); \
+        g_autofree gchar *ane_message = \
+            g_strdup_printf ("Expected no signal emissions, but saw %" G_GSIZE_FORMAT "\n:%s", \
+                             mws_signal_logger_get_n_emissions (self), \
+                             ane_list); \
+        g_assertion_message (G_LOG_DOMAIN, __FILE__, __LINE__, G_STRFUNC, \
+                             ane_message); \
+    }\
+  } G_STMT_END
```

Comment:
I assume you want this a macro for better located logging, right? I wonder if it should go all in its function except for the `g_assertion_message`, so we reduce the code inside the macro as it can easily get messier.

Reply:
You’re right that the macro exists so the logging has the right line number and function number for error messages. I deliberately didn’t put more of the implementation into a helper function, as that would mean adding another public function which would be totally useless to anyone who wasn’t implementing their own `mws_signal_logger_assert_no_emissions()` macro (due to having a very specific error message, etc.).

Reply:
> as that would mean adding another public function which would be totally useless to anyone who wasn’t implementing their own mws_signal_logger_assert_no_emissions() macro

Sometimes it's worth declaring a function just for self-documented code. But it's fine.